add reverse

cryptix opened the pull request:
this enables getting the last message first and iterating backwards.

I disabled the tests for offset v1 and added a deprecation warning. We only used this format at the beginning of our ssb work. It Still works but I didn't add reverse to it so the tests fail.

cc @keks

cryptix commented:
@keks would love a shallow review from you or maybe just some suggestions for additional test cases?

keks commented on mem/log.go:
```diff
@@ -124,7 +125,20 @@ func (log *memlog) Append(v interface{}) (margaret.Seq, error) {
+		wait: make(chan struct{}),
+	}
+
+	// scroll to prev
```
So we are looking for the _second_ last here? Why don't we keep track of that in `log.tail`? That way we don't have to seek through the entire thing every time.

I mean it's only the memory-backed implementation, but still...

cryptix replied:
yup, this was a brainfart.

i already added the backlink pointer `prev` to the elements for seeking backwards.

keks commented on mem/qry.go:
```diff
@@ -106,8 +120,23 @@ func (qry *memlogQuery) Next(ctx context.Context) (interface{}, error) {
 	defer qry.log.l.Unlock()
 
+	if qry.reverse {
+		if qry.live {
```
Maybe this check should be done when applying the option to the query? Both in Live and Reverse.

cryptix replied:
I tried that but only got it working for the normal, not Push mode.

keks commented on multilog/badger/qry.go:
```diff
@@ -58,6 +59,23 @@ func (qry *query) Lte(s margaret.Seq) error {
 }
 
+func (qry *query) Reverse(yes bool) error {
+	qry.reverse = yes
```
add check here that live is not set, and do a check in Live that reverse is not set

cryptix replied:
done but needs to happen in the logs `Query()` since all the specs have to be applied first.

keks commented on multilog/sink.go:
```diff
@@ -70,6 +70,8 @@ func (slog *sinkLog) Pour(ctx context.Context, v interface{}) error {
 }
 
 // Close closes the root-log..? seems wrong
+// rational here beeing having multiple mlogs on one root (?)
```
The code that's commented out doesn't close the root log but the multilog, which seems like a reasonable thing to do. But yeah, not sure when this function get's called, and whether the multilog is being read afterwards. Maybe have something like a "multiple-open tracker" that tracks all the users of the multilog and closes it if all users closed it? Not saying this should happen now, but let's at least put a TODO here that we need to figure out how to proceed correctly.

keks commented on offset2/qry.go:
```diff
@@ -77,6 +78,29 @@ func (qry *offsetQuery) SeqWrap(wrap bool) error {
 
+func (qry *offsetQuery) Reverse(yes bool) error {
+	qry.reverse = yes
+	if yes {
```
check here that live is not set, and check in Live that reverse is not set

keks commented on offset/qry.go:
```diff
@@ -75,6 +76,10 @@ func (qry *offsetQuery) SeqWrap(wrap bool) error {
 	qry.seqWrap = wrap
 	return nil
 }
+func (qry *offsetQuery) Reverse(yes bool) error {
```
Hmm, maybe better remove this? The way it is now looks like it's supported. Or, if we want it to still implement the interface, just return a "not implemented" error

keks commented on test/main.go:
```diff
@@ -13,6 +13,6 @@ func LogTest(f NewLogFunc) func(*testing.T) {
 		t.Run("Simple", LogTestSimple(f))
 		t.Run("Concurrent", LogTestConcurrent(f))
-		// t.Run("Copy", LogTestCopy(f))
+		// t.Run("Copy", LogTestCopy(f)) // TODO: what's this supposed to be?
```
Don't remember, feel free to delete 😅

cryptix commented:
updated all of that. Also added tests for rev+gt, which are failing right now.

I think the main blocker is that options need to be cleaned up. I'd argue that we can't do validation for the whole query inside them (since that depends on order and the breaking ones might not have been applied yet).

Also the way that GT(e) just seeks forward breaks the way I implemented reverse but that's a nother cup of 🍵 from the above but the ball is in the same corner of the field.

keks commented:
> I think the main blocker is that options need to be cleaned up.

Agreed.

> I'd argue that we can't do validation for the whole query inside them (since that depends on order and the breaking ones might not have been applied yet).

I don't think so. We can check that `qry.reverse` is `false` when `qry.Live(true)` gets called, and vice versa. That would mean that we have to put the check in both Live and Reverse, but IMHO that is the clean way to do it, and that's why the Options return errors in the first place.

Currently `"offset2".log.Query` doesn't do anything except applying the options and I really like that the options are self-contained.

cryptix commented:
> that's why the Options return errors in the first place

there are things that are self-contained to the option, like only accepting a limit of a certain size. Doing it that way means having to check it in two places and having to keep that in sync as the code changes.

> Currently "offset2".log.Query doesn't do anything except applying the options and I really like that the options are self-contained.

If you look at the GTE/LTE + Reverse case, it gets very messy to do the setup in the options since you need to respect this interplay with other options which might be called before or after another one, so the init/setup code gets much simpler if you try to do it once everything was applied,

keks commented:
I know that it would be a little bit simpler if we do it afterwards, it just has a much saner structure if we do it in the option itself. An option is a thing that changes the query state and then returns an error. If a change is illegal, it should return an error immediately, instead of saying "all good" and having to verify later whether really everything is ok.

We could try doing the check in a function and using that in both places?

PragmaticCypher commented on mem/log.go:
```diff
@@ -117,15 +119,26 @@ func (log *memlog) Query(specs ...margaret.QuerySpec) (luigi.Source, error) {
 	}
 
+	if qry.reverse && qry.live {
+		return nil, errors.Errorf("memlog: can't do reverse and live")
```
This error message feels vague for someone unfamiliar with the code.

"can't do reverse and live (what?)"

How about "memlog: reverse and live are incompatible query parameters"?

PragmaticCypher commented on multilog/badger/qry.go:
```diff
@@ -133,7 +154,11 @@ func (qry *query) Next(ctx context.Context) (interface{}, error) {
+	if qry.reverse {
+		defer func() { qry.nextSeq-- }()
+	} else {
+		defer func() { qry.nextSeq++ }()
```
This looks like an abuse of the defer function :|  maybe add it inline at the end of the function?

PragmaticCypher commented on offset2/log.go:
```diff
@@ -420,6 +420,10 @@ func (log *offsetLog) Query(specs ...margaret.QuerySpec) (luigi.Source, error) {
 	}
 
+	if qry.reverse && qry.live {
+		return nil, errors.Errorf("offset2: can't do reverse and live")
```
Same as mentioned before, I think the message could be clearer here :)

PragmaticCypher commented on mem/qry.go:
```diff
@@ -97,6 +98,19 @@ func (qry *memlogQuery) Live(live bool) error {
 	return nil
 }
 
+func (qry *memlogQuery) SeqWrap(wrap bool) error {
```
Just curious, what does SeqWrap do?

keks commented:
It makes the the calls to Next return a type that let's you access both the value and the sequence number of that value in the log.

Thanks for the review!

Am 10. Juli 2019 11:34:48 MESZ schrieb PragmaticCypher <notifications@github.com>:
>PragmaticCypher commented on this pull request.
>
>
>
>> @@ -97,6 +98,19 @@ func (qry *memlogQuery) Live(live bool) error {
> 	return nil
> }
> 
>+func (qry *memlogQuery) SeqWrap(wrap bool) error {
>
>Just curious, what does SeqWrap do?

-- 
Sent from mobile.
